### PR TITLE
Arduino python scripts

### PR DESCRIPTION
**Pull request by ddavis-2015**
@tensorflow/micro

Arduino library creation for CI build.  Part 1 of PR for Issue #3

BUG=https://github.com/tensorflow/tflite-micro/issues/256

### CODE REVIEW

**Review by advaitjain — CHANGES_REQUESTED** [review]
let's split this PR into two PRs:

1. copy over all the files from tflite-micro that will live in the current repo once this work is complete
2. PR with only the new code that you have written for the sync -- I'm assuming that is the manifest.ini and the create_tflm_arduino.py

Also, let's put the files in micro/tools/make, micro/tools/ci_build and micro/tools/project_generation into a top level ci/ or scripts/ directory.

I don't have a preference for the name of the directory, but would like to have a separate top-level directory with all the code needed to handle the sync and Arduino library generation (rather than being nested within the src/tensorflow tree).

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+
+  Args:
+    file_path: The full path to the input file
+    scripts: A list of strings, each containing a single SED script
+    args: a string containing all the SED arguments | None
+    is_dry_run: if True, do not execute any commands
+  """
+  cmd = "sed"
+  if args is not None and args != "":
+    cmd += f" {args}"
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:77]: combining string concatenation with f-strings seems to defeat the purpose of f-strings.

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+  for scr in scripts:
+    cmd += f" -E {scr}"
+  cmd += f" < {file_path!s}"
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:80]: I think we need to enforce a few conditions on the parameters to this function:

args: must not be an empty string
scripts: must not be an empty list

The we can replace all of these with:

```py
cmd = f"sed {args} -E {" -E ".join(scripts)} {str(file_path)}"
```

Note that `{file_path!s}` is a bit too fancy for me. Let's use the more explicit `str(file_path)`

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+  print(f"Running command: {cmd}")
+  if not is_dry_run:
+    result = subprocess.run(cmd,
+                            shell=True,
+                            check=True,
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.PIPE)
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:87]: let's also capture the error code and raise an exception if the command did not succeed.

Similar to:
https://github.com/tensorflow/tflite-micro/blob/43265614fe010c677c447c1269500478b49e72dd/tensorflow/lite/micro/tools/project_generation/create_tflm_tree.py#L53-L57

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+#   sed -E 's/utility\.h/utility/g'
+# patch third_party/kissfft/kiss_fft.h
+#   sed -E 's@#include <string.h>@//#include <string.h>
+# 	 /* Patched by helper_functions.inc for Arduino compatibility */@g'
+# run fix_arduino_subfolders.py {outdir}/tflm_arduino
+# remove all empty directories in {outdir}/tflm_arduino tree
+# create ZIP file using shutil.make_archive()
+
+
+def RunSedScripts(file_path: Path,
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:61]: naming nit: _run_sed_scripts

https://google.github.io/styleguide/pyguide.html#316-naming

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+    result = subprocess.run(cmd,
+                            shell=True,
+                            check=True,
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.PIPE)
+    print(f"Saving output to: {file_path!s}")
+    file_path.write_bytes(result.stdout)
+
+
+def RemoveDirectories(paths: List[Path], is_dry_run: bool = True) -> None:
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:92]: _remove_directories

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+  """
+  for dir_path in paths:
+    print(f"Removing directory tree {dir_path!s}")
+    if dir_path.exists() and not is_dry_run:
+      shutil.rmtree(dir_path)
+
+
+def RemoveEmptyDirectories(paths: Iterable[Path],
+                           root: Path,
+                           is_dry_run: bool = True) -> None:
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:108]: not clear to me why in the overal generation process we will have empty directories. Let's talk about the need for this function.

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+  """
+  Run a python script with specified arguments
+
+  Args:
+    path_to_script: The full path to the Python script
+    args: a string containing all the script arguments
+    is_dry_run: if True, do not execute any commands
+  """
+  cmd = f"python3 {path_to_script!s}"
+  cmd += f" {args}"
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:143]:
```py
cmd = f"python3 {path_to_script} {args}"
```

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+  print(f"Running command: {cmd}")
+  if not is_dry_run:
+    _ = subprocess.run(cmd,
+                       shell=True,
+                       check=True,
+                       stdout=subprocess.PIPE,
+                       stderr=subprocess.PIPE)
+
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:151]: please catch errors in the scripts and raise exceptions.

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+      continue
+    parent_paths.add(dir_path.parent)
+    print(f"Removing empty directory {dir_path!s}")
+    if not is_dry_run:
+      dir_path.rmdir()
+  if len(parent_paths) > 0:
+    RemoveEmptyDirectories(parent_paths, root=root, is_dry_run=is_dry_run)
+
+
+def RunPythonScript(path_to_script: Path,
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:131]: doesn't look like we need `path_to_script` to be of type Path. str seems sufficient here.

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+
+def CreateDirectories(paths: List[Path], is_dry_run: bool = True) -> None:
+  """
+  Create directory tree(s) given list of pathnames
+
+  Args:
+    paths: A list of Path objects
+    is_dry_run: if True, do not execute any commands
+  """
+  dir_path: Path
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:161]: remove.

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+  """
+  Copy files given list of source and destination Path tuples
+
+  Args:
+    paths: A list of tuples of Path objects.
+    Each tuple is of the form (source, destination)
+    is_dry_run: if True, do not execute any commands
+  """
+  dir_path: Tuple[Path, Path]
+  for dir_path in paths:
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:179]: dir_path is not the correct name for this variable. my preference would be to have two lists, src and dest.

we can have a precondition to check that the two lists are of the same length.

File: src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py [hunk 12]
```diff
@@ -0,0 +1,548 @@
...
+              is_dry_run: bool = True) -> None:
+  """
+  Copy files given list of source and destination Path tuples
+
+  Args:
+    paths: A list of tuples of Path objects.
+    Each tuple is of the form (source, destination)
+    is_dry_run: if True, do not execute any commands
+  """
+  dir_path: Tuple[Path, Path]
```
Comment (advaitjain) [inline src/tensorflow/lite/micro/tools/project_generation/create_tflm_arduino.py:178]: remove, not necessary.

**Review by advaitjain — CHANGES_REQUESTED** [review]
I haven't reviewed all of the code yet. happy to chat more tomorrow.

File: scripts/create_tflm_arduino.py [hunk 13]
```diff
@@ -0,0 +1,504 @@
...
+    # parse manifest file
+    manifest = self._ParseManifest()
+    self.add_list: List[Path] = manifest[self.Manifest.ADD]
+    self.remove_list: List[Path] = manifest[self.Manifest.REMOVE]
+    self.special_repo_list: List[Tuple[Path, Path]] = manifest[
+        self.Manifest.SPECIAL_REPO]
+    self.special_base_list: List[Tuple[Path, Path]] = manifest[
+        self.Manifest.SPECIAL_BASE]
+    self.patch_sed_list: List[Tuple[List[Path], List[str]]] = manifest[
+        self.Manifest.PATCH_SED]
```
Comment (advaitjain) [inline scripts/create_tflm_arduino.py:195]: why not assign self.add_list etc. in the _parse_manifest function? Then we won't need to create and think about the intermediate dict struct.

File: scripts/create_tflm_arduino.py [hunk 14]
```diff
@@ -0,0 +1,504 @@
...
+
+  def _GenerateRepoPathsRelative(
+      self, generate_transform_paths: bool) -> List[Tuple[Path, Path]]:
+
+    relative_path_pairs: List[Tuple[Path, Path]] = []
+    if generate_transform_paths:
+      # future: download (url=) sections using transform_suffixes= list
+      # future: apply suffixes_mapping=
+      # for now just return empty list
+      pass
```
Comment (advaitjain) [inline scripts/create_tflm_arduino.py:449]: what is this for?

File: scripts/create_tflm_arduino.py [hunk 15]
```diff
@@ -0,0 +1,504 @@
...
+                              is_dry_run=self.is_dry_run)
+
+  #
+  # public methods
+  #
+
+  def CreateZip(self) -> None:
+    """
+    Execute all steps to create TFLM Arduino ZIP file
+    """
```
Comment (advaitjain) [inline scripts/create_tflm_arduino.py:476]: Let's discuss what we want the script output to be. I would prefer this script to change the repository in-place with updated files from tflite-micro unless that would become quite a bit more complex.

**Review by advaitjain — CHANGES_REQUESTED** [review]

File: scripts/MANIFEST.ini [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+#
+[DEFAULT]
+# currently empty
+
+#
+# Required section.
+#
+# The <files> key is required, however the value is allowed to be empty.
+# The <files> values are the files in the repository that will be copied
+# without modification to the output directory.
```
[the remark this answers is not in the document]
Comment (ddavis-2015) [reply scripts/MANIFEST.ini:47]: This will be in the README.md in the scripts directory

**Review by advaitjain — APPROVED** [review]

**Pull request merged**